mcm/json_layer/request.py

```diff
@@ -1872,9 +1872,46 @@ def get_stats(self, forced=False):
                                                   page=0,
                                                   limit=-1,
                                                   options={'key': prepid})
+        # easier to access this way below
+        stats_workflows_dict = dict([(stats_workflow.get('RequestName'), stats_workflow) for stats_workflow in stats_workflows])
+
         mcm_reqmgr_list = self.get_attribute('reqmgr_name')
         mcm_reqmgr_name_list = [x['name'] for x in mcm_reqmgr_list]
-        stats_reqmgr_name_list = [stats_wf['RequestName'] for stats_wf in stats_workflows]
+
+        # Attempt to pick the latest injected workflow using the records available in McM
+        # After an injection process finishes, only the latest workflow name is available on the `reqmgr_name`
+        # attribute with no content.
+        # Try to fetch it from Stats2 and pick the date from the `new` transition.
+        #
+        # Note this process is not idempotent. In case the request has been updated with Stats2
+        # data before, nothing will be filtered. Such behavior is acceptable.
+        workflows_to_consider = [
+            stats_workflows_dict.get(mcm_reqmgr_name)
+            for mcm_reqmgr_name in mcm_reqmgr_name_list
+            if stats_workflows_dict.get(mcm_reqmgr_name)
+        ]
+        self.logger.info("Workflows to consider: %s", [doc.get('RequestName') for doc in workflows_to_consider])
+        try:
+            reference_times_of_new = min([
+                stats_doc.get('RequestTransition')[0].get('UpdateTime')
+                for stats_doc in workflows_to_consider
+            ])
+        except:
+            # The workflows in mcm are too fresh and have not yet been registered in stats. update will have to wait.
+            traceback.print_exc()
+            return False
+
+        # remove anything that predates the exiting workflows registered to mcm
+        to_remove = [
+            reqmgr_name
+            for (reqmgr_name, content) in stats_workflows_dict.iteritems()
+            if content.get('RequestTransition') and
+            content['RequestTransition'][0].get('UpdateTime') < reference_times_of_new
+        ]
+        popped = [stats_workflows_dict.pop(reqmgr_name) for reqmgr_name in to_remove]
+        self.logger.info("The following workflows may not be related to the last injection: %s", [doc.get('RequestName') for doc in popped])
+
+        stats_reqmgr_name_list = stats_workflows_dict.keys()
```

Comment:
We don't actually need to remove the workflows from `stats_workflows_dict`, so this can be simplified:
```suggestion
        stats_reqmgr_name_list = [
            reqmgr_name
            for reqmgr_name, content in stats_workflows_dict.iteritems()
            if not content.get('RequestTransition') or
                content['RequestTransition'][0].get('UpdateTime') >= reference_times_of_new
        ]
        self.logger.info("The following workflows may not be related to the last injection: %s", set(stats_workflows_dict.keys()) - set(stats_reqmgr_name_list))
```
(explicit conversion to `set` is not needed in Python 3)

Reply:
I tweaked the suggestion to check some assumptions about the `elements of stats_workflow_dict`